src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -461,6 +472,138 @@ bool ExpressionAnalyzer::makeAggregateDescriptions(ActionsDAGPtr & actions)
+                // Ignore collation for now.
+                window_description.order_by.push_back(
+                    SortColumnDescription(
+                        order_by_element.children.front()->getColumnName(),
+                        order_by_element.direction,
+                        order_by_element.nulls_direction));
+            }
+        }
+
+        window_description.full_sort_description = window_description.partition_by;
+        window_description.full_sort_description.insert(
+            window_description.full_sort_description.end(),
+            window_description.order_by.begin(),
+            window_description.order_by.end());
+
+        WindowFunctionDescription window_function;
+        window_function.window_name = window_description.window_name;
+        window_function.function_node = function_node;
+        window_function.column_name
+            = window_function.function_node->getColumnName();
+        window_function.function_parameters
+            = window_function.function_node->parameters
+                ? getAggregateFunctionParametersArray(
+                    window_function.function_node->parameters)
+                : Array();
+
+        // Requiring a constant reference to a shared pointer to non-const AST
+        // doesn't really look sane, but the visitor does indeed require it.
+        // Hence we clone the node (not very sane either, I know).
+        getRootActionsNoMakeSet(window_function.function_node->clone(),
+            true, actions);
+
+        const ASTs & arguments
+            = window_function.function_node->arguments->children;
+        window_function.argument_types.resize(arguments.size());
+        window_function.argument_names.resize(arguments.size());
+        const auto & index = actions->getIndex();
+        for (size_t i = 0; i < arguments.size(); ++i)
+        {
+            const std::string & name = arguments[i]->getColumnName();
+
+            auto it = index.find(name);
+            if (it == index.end())
+            {
+                throw Exception(ErrorCodes::UNKNOWN_IDENTIFIER,
+                    "Unknown identifier '{}' in window function '{}'",
+                    name, window_function.function_node->formatForErrorMessage());
+            }
+
+            window_function.argument_types[i] = (*it)->result_type;
+            window_function.argument_names[i] = name;
+        }
+
+        AggregateFunctionProperties properties;
+        window_function.aggregate_function
+            = AggregateFunctionFactory::instance().get(
+                window_function.function_node->name,
+                window_function.argument_types,
+                window_function.function_parameters, properties);
+
+        window_function_descriptions.push_back(window_function);
+
+        if (auto it = window_descriptions.find(window_description.window_name);
+            it != window_descriptions.end())
+        {
+            assert(it->second.full_sort_description
```

Comment:
This check looks a little bit clumsy ...
I would rather group by sort `full_sort_description` initially.

Also, I suppose it may be reasonable group windows by `partition_by`, cause we can start common sorting for a window by it, and then add partial sorting by `full_sort_description` for each window function separately.

Reply:
I think I'll try to write the complete window matching (group the windows declared ad hoc with same parameters) + window declaration grammar, to figure this out.

Reply:
Some windows are just totally identical, and then to some windows we can apply the optimization you describe (group them so that they have a common sorting prefix). I have both these points here in the "rough edges" list: https://github.com/ClickHouse/ClickHouse/issues/18097